### DISCUSSION TITLE
How to run linter on code that depends on packages outside the analyzed elements.

### DISCUSSION THREAD

**Discussion by harshit2202** [golangci/golangci-lint#4818] [Q&A]
I am trying to run linter on one module(let's name this `module-1`) in our mono-repo setup, which depends on another module(`module-2`). And it is failing stating typecheck error for all the imports that are from the module-2 in module-1.

On investigating this I found that this is already mentioned in the https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors

```Ensure you are not running an analysis on code that depends on files/packages outside the scope of the analyzed elements.```

Is there any workaround for this? I would appreciate any help.

Current file structure is
```
- main-github-repo
  - module-1
    - go.mod 
    - .golangci.yaml
    - ....
  - module-2
```

In the `module-1/go.mod`, we import the dependency `module-2` using
```replace github.com/main-github-repo-/module-1 v1.0.0 => ../module-1```

**Comment by ldez** [comment]
hello,

a `replace` is not a problem by itself, it works.

You need to run golangci-lint inside a module (ex: the directory `module-1)`, not inside the root (`main-github-repo`).

**Reply by harshit2202** [reply]
I am running it inside `module-1` directory and yet it is giving the same error.

**Reply by ldez** [reply]
I created a simple project:

```
main-github-repo
├── module-1
│ ├── go.mod
│ └── main.go
└── module-2
    ├── foo.go
    └── go.mod
```


<details>
<summary>Project content</summary>

- `module-1/go.mod`
```go
module github.com/golangci/moduleone

go 1.21.0

require github.com/golangci/moduletwo v0.0.0-00010101000000-000000000000

replace github.com/golangci/moduletwo => ../module-2

```

- `module-1/main.go`
```go
package main

import (
	"errors"

	"github.com/golangci/moduletwo"
)

func main() {
    moduletwo.Foo()
    Bar()
}

func Bar() error {
	return errors.New("OOPS")
}

```

- `module-2/go.mod`
```go
module github.com/golangci/moduletwo

go 1.21.0

```

- `module-2/main.go`
```go
package moduletwo

import "fmt"

func Foo() {
    fmt.Println("Foo")
}

```

</details>

```console
$ cd module-1

$ golangci-lint run
main.go:11:8: Error return value is not checked (errcheck)
    Bar()
       ^
```

As you can see it works as expected.

Your problem is maybe a missing call to `go mod tidy` inside your modules.

**Reply by harshit2202** [reply]
> As you can see it works as expected.

By expected, you mean giving error? If yes, is there any solution for this?

This is already mentioned in https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors, but is there any way to run golangci-lint for such module structures as well(where module-1 is dependent on module-2) ?

**Reply by ldez** [reply]
> By expected, you mean giving error?

No, in my example it's an `errcheck` report, not a `typecheck` error.
I added a linting issue for the example.

Click ["Project content"](https://github.com/golangci/golangci-lint/discussions/4818#discussioncomment-9775071) to see the project source files.
